### PR TITLE
Fix pql double quote checker exception

### PR DESCRIPTION
**Pull request by walterddr**
Currently if there's a double quoted PQL syntax ingestion transform such as
```
  "ingestionConfig": {
    "transformConfigs": [
      {
        "columnName": "created_at_timestamp",
        "transformFunction": "fromDateTime(created_at, \"yyyy-MM-dd''T''HH:mm:ss''Z''\")"
      }
    ]
  },
```
It will fail because the double quoted literal is considered as quoted identifier in Calcite SQL parser. 
This is not a problem during SQL execution because the column will not be found. but during ingestion we dont have a proper schema validation thus exception will not be meaningful.

Propose to wrap exception in InbuiltFunctionEvaluator that executes any ScalarFuntion in FunctionRegistry and provide extra context information about the argument that passed in. 

Before:
```
java.lang.IllegalStateException: Caught exception while invoking method: public static long org.apache.pinot.common.function.scalar.DateTimeFunctions.fromDateTime(java.lang.String,java.lang.String) with arguments: [2021-04-20T16:41:00-0700, null]
	at org.apache.pinot.common.function.FunctionInvoker.invoke(FunctionInvoker.java:131)
	at 
...
Caused by: java.lang.reflect.InvocationTargetException
	at jdk.internal.reflect.GeneratedMethodAccessor33.invoke(Unknown Source)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.pinot.common.function.FunctionInvoker.invoke(FunctionInvoker.java:128)
	... 8 more
Caused by: java.lang.IllegalArgumentException: Invalid pattern specification
	at org.joda.time.format.DateTimeFormat.createFormatterForPattern(DateTimeFormat.java:682)
	at org.joda.time.format.DateTimeFormat.forPattern(DateTimeFormat.java:177)
```

Now:
```
... [Extra wrapper exception with column info]
java.lang.RuntimeException: Caught exception while executing function: fromDateTime[reverse['2020-01-01T00:00:00Z'], invalid_identifier]
	at org.apache.pinot.segment.local.function.InbuiltFunctionEvaluator$FunctionExecutionNode.execute(InbuiltFunctionEvaluator.java:126)
	at org.apache.pinot.segment.local.function.InbuiltFunctionEvaluator.evaluate(InbuiltFunctionEvaluator.java:88)
... [Original Exception]
Caused by: java.lang.IllegalStateException: Caught exception while invoking method: public static long org.apache.pinot.common.function.scalar.DateTimeFunctions.fromDateTime(java.lang.String,java.lang.String) with arguments: [2018-01-01T11:00:00Z, null]
  	at org.apache....
...
```

### CODE REVIEW

File: pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java [hunk 1]
```diff
@@ -18,18 +18,28 @@
...
 
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertTrue;
+import static org.testng.AssertJUnit.fail;
 
 
 public class InbuiltFunctionEvaluatorTest {
 
+  @BeforeClass
+  public void setUp()
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java:37]: This function is specific to one test, so I suggest keeping it within the test

File: pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java [hunk 2]
```diff
@@ -120,10 +128,27 @@ public void testStateSharedBetweenRowsForExecution()
...
+      assertTrue(e instanceof RuntimeException);
+      assertTrue(e.getCause() instanceof IllegalStateException);
+    }
+  }
+
   private static class MyFunc {
     String _baseString = "";
 
     String appendToStringAndReturn(String addedString) {
+      Preconditions.checkNotNull(addedString);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java:151]: Seems unrelated?

File: pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java [hunk 3]
```diff
@@ -120,10 +128,27 @@ public void testStateSharedBetweenRowsForExecution()
     assertEquals(evaluator.evaluate(row), "test test test ");
   }
 
+  @Test
+  public void testExceptionDuringInbuiltFunctionEvaluator()
+      throws Exception {
+    String expression = "fromDateTime('2020-01-01T00:00:00Z', \"invalid_identifier\")";
+    InbuiltFunctionEvaluator evaluator = new InbuiltFunctionEvaluator(expression);
+    assertEquals(1, evaluator.getArguments().size());
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java:136]: First argument is actual, second argument is expected. It will be easier for debugging when it fails
```suggestion
    assertEquals(evaluator.getArguments().size(), 1);
```

File: pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java [hunk 4]
```diff
@@ -120,10 +128,27 @@ public void testStateSharedBetweenRowsForExecution()
...
+    GenericRow row = new GenericRow();
+    try {
+      evaluator.evaluate(row);
+      fail();
+    } catch (Exception e) {
+      assertTrue(e instanceof RuntimeException);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java:142]: Should we check the actual exception message to verify the expected behavior?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java [hunk 5]
```diff
@@ -112,22 +113,39 @@ public Object evaluate(Object[] values) {
...
+            "Caught exception while execute function: " + _functionInvoker.getMethod().getName()
+                + " with argument nodes: " + Arrays.toString(_argumentNodes), e);
       }
-      _functionInvoker.convertTypes(_arguments);
-      return _functionInvoker.invoke(_arguments);
+    }
+
+    @Override
+    public String toString() {
+      return String.format("[FunctionExecutionNode[functionName:%s]]", _functionInvoker.getMethod().getName());
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java:148]: We should also include the `_argumentNodes`. I'd suggest returning the format of `function(column, 'const')` for readability and easier debugging. You may refer to the `ExpressionContext` class on how we serialize the expression.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java [hunk 6]
```diff
@@ -112,22 +113,39 @@ public Object evaluate(Object[] values) {
...
+      try {
+        int numArguments = _argumentNodes.length;
+        for (int i = 0; i < numArguments; i++) {
+          _arguments[i] = _argumentNodes[i].execute(row);
+        }
+        _functionInvoker.convertTypes(_arguments);
+        return _functionInvoker.invoke(_arguments);
+      } catch (Exception e) {
+        throw new RuntimeException(
+            "Caught exception while execute function: " + _functionInvoker.getMethod().getName()
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java:125]: (nit)
```suggestion
            "Caught exception while executing function: " + _functionInvoker.getMethod().getName()
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java [hunk 7]
```diff
@@ -112,22 +113,39 @@ public Object evaluate(Object[] values) {
...
+      try {
+        int numArguments = _argumentNodes.length;
+        for (int i = 0; i < numArguments; i++) {
+          _arguments[i] = _argumentNodes[i].execute(values);
+        }
+        _functionInvoker.convertTypes(_arguments);
+        return _functionInvoker.invoke(_arguments);
+      } catch (Exception e) {
+        throw new RuntimeException(
+            "Caught exception while execute function: " + _functionInvoker.getMethod().getName()
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java:141]: Same

**Review by Jackie-Jiang — COMMENTED** [review]
Mostly good

File: pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java (hunk 3 above)
Comment (walterddr) [reply pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java:136]: geeze.. this is revert convention from junit!!!!
http://junit.sourceforge.net/javadoc/org/junit/Assert.html
https://www.javadoc.io/doc/org.testng/testng/6.8.17/org/testng/Assert.html

File: pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java (hunk 2 above)
Comment (walterddr) [reply pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java:151]: yeah sorry I was planning to use MyFunc, will revert this

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java [hunk 8]
```diff
@@ -112,22 +113,37 @@ public Object evaluate(Object[] values) {
...
+      try {
+        int numArguments = _argumentNodes.length;
+        for (int i = 0; i < numArguments; i++) {
+          _arguments[i] = _argumentNodes[i].execute(row);
+        }
+        _functionInvoker.convertTypes(_arguments);
+        return _functionInvoker.invoke(_arguments);
+      } catch (Exception e) {
+        throw new RuntimeException(
+            "Caught exception while executing function: " + this, e);
```
Comment (walterddr) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java:125]:
```suggestion
        throw new RuntimeException("Caught exception while executing function: " + this, e);
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java [hunk 9]
```diff
@@ -112,22 +113,37 @@ public Object evaluate(Object[] values) {
...
+      try {
+        int numArguments = _argumentNodes.length;
+        for (int i = 0; i < numArguments; i++) {
+          _arguments[i] = _argumentNodes[i].execute(values);
+        }
+        _functionInvoker.convertTypes(_arguments);
+        return _functionInvoker.invoke(_arguments);
+      } catch (Exception e) {
+        throw new RuntimeException(
+            "Caught exception while executing function: " + this, e);
```
Comment (walterddr) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/InbuiltFunctionEvaluator.java:140]:
```suggestion
        throw new RuntimeException("Caught exception while executing function: " + this, e);
```

File: pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java [hunk 10]
```diff
@@ -26,6 +26,7 @@
 
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertTrue;
+import static org.testng.AssertJUnit.fail;
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java:29]: (nit)
```suggestion
import static org.testng.Assert.fail;
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM. Thanks for fixing it

File: pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java (hunk 10 above)
Comment (walterddr) [reply pinot-core/src/test/java/org/apache/pinot/core/data/function/InbuiltFunctionEvaluatorTest.java]: sounds good although we should really do these automatically. not a great use of any committer's time :-)
